// maishu-jueying-core/out/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentContainer = exports.PageBody = exports.PageFooter = exports.PageHeader = exports.Page = exports.componentTypes = exports.registerComponent = exports.parseComponentData = void 0;
var parse_component_data_1 = require("./parse-component-data");
Object.defineProperty(exports, "parseComponentData", { enumerable: true, get: function () { return parse_component_data_1.parseComponentData; } });
var register_1 = require("./register");
Object.defineProperty(exports, "registerComponent", { enumerable: true, get: function () { return register_1.registerComponent; } });
Object.defineProperty(exports, "componentTypes", { enumerable: true, get: function () { return register_1.componentTypes; } });
var index_1 = require("./components/index");
Object.defineProperty(exports, "Page", { enumerable: true, get: function () { return index_1.Page; } });
Object.defineProperty(exports, "PageHeader", { enumerable: true, get: function () { return index_1.PageHeader; } });
Object.defineProperty(exports, "PageFooter", { enumerable: true, get: function () { return index_1.PageFooter; } });
Object.defineProperty(exports, "PageBody", { enumerable: true, get: function () { return index_1.PageBody; } });
var component_container_1 = require("./component-container");
Object.defineProperty(exports, "ComponentContainer", { enumerable: true, get: function () { return component_container_1.ComponentContainer; } });
require("./style");
