// maishu-jueying-core/out/style.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFixed = exports.isVisible = void 0;
const components_1 = require("./components");
let elementId = "maishu-jueying-core-style";
if (!document.getElementById(elementId) && document.head != null) {
    let element = document.createElement('style');
    element.type = 'text/css';
    element.id = elementId;
    document.head.appendChild(element);
    element.innerHTML = `
    .${components_1.Page.className} {
        width   : 100%;
        height  : 100%;
        position: absolute;
    }
    
    .${components_1.Page.className} .${components_1.PageBody.className} {
        overflow-y: auto;
        overflow-x: hidden;
        height  : 100%;
    }
    
    .${components_1.Page.className} .${components_1.PageBody.className}::-webkit-scrollbar-track-piece {
        background-color: #fff;
    }
    
    .${components_1.Page.className} .${components_1.PageBody.className}::-webkit-scrollbar {
        width: 4px;
    }
    
    .${components_1.Page.className} .${components_1.PageBody.className}::-webkit-scrollbar-thumb {
        background: #999;
    }
    
    .${components_1.Page.className} .${components_1.PageHeader.className} {
        position: absolute;
        top     : 0px;
        width   : 100%;
    }
    
    .${components_1.Page.className} .${components_1.PageFooter.className} {
        position: absolute;
        bottom  : 0px;
        width   : 100%;
        margin  : 0;
    }
    `;
}
function isVisible(style) {
    if (!style.display)
        return true;
    return style.display != "none";
}
exports.isVisible = isVisible;
function isFixed(style) {
    return style.position == "fixed" || style.position == "absolute";
}
exports.isFixed = isFixed;
