// maishu-jueying/out-es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ComponentDataHandler = exports.classNames = exports.TextInput = exports.PropEditor = exports.DesignerContext = exports.PageDesigner = exports.EditorPanel = exports.Component = exports.groupDisplayNames = void 0;

var common_1 = require("./common");

Object.defineProperty(exports, "groupDisplayNames", {
  enumerable: true,
  get: function get() {
    return common_1.groupDisplayNames;
  }
});

var component_1 = require("./component");

Object.defineProperty(exports, "Component", {
  enumerable: true,
  get: function get() {
    return component_1.Component;
  }
});

var editor_panel_1 = require("./editor-panel");

Object.defineProperty(exports, "EditorPanel", {
  enumerable: true,
  get: function get() {
    return editor_panel_1.EditorPanel;
  }
});

var page_designer_1 = require("./page-designer");

Object.defineProperty(exports, "PageDesigner", {
  enumerable: true,
  get: function get() {
    return page_designer_1.PageDesigner;
  }
});
Object.defineProperty(exports, "DesignerContext", {
  enumerable: true,
  get: function get() {
    return page_designer_1.DesignerContext;
  }
});

var prop_editor_1 = require("./prop-editor");

Object.defineProperty(exports, "PropEditor", {
  enumerable: true,
  get: function get() {
    return prop_editor_1.PropEditor;
  }
});
Object.defineProperty(exports, "TextInput", {
  enumerable: true,
  get: function get() {
    return prop_editor_1.TextInput;
  }
});

var style_1 = require("./style");

Object.defineProperty(exports, "classNames", {
  enumerable: true,
  get: function get() {
    return style_1.classNames;
  }
});

var component_data_handler_1 = require("./component-data-handler");

Object.defineProperty(exports, "ComponentDataHandler", {
  enumerable: true,
  get: function get() {
    return component_data_handler_1.ComponentDataHandler;
  }
});
