// maishu-nws-mvc/out/api-controller.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAPIControllerType = void 0;
const controller_1 = require("./controller");
const attributes_1 = require("./attributes");
const maishu_node_web_server_1 = require("maishu-node-web-server");
function createAPIControllerType(getActionInfos, serverContext) {
    let APIControllerType = class APIController extends controller_1.Controller {
        async list() {
            let actionInfos = getActionInfos();
            let r = actionInfos.map(o => ({
                path: o.actionPath,
                controller: o.controllerType.name,
                action: o.memberName,
                filePath: o.controllerPhysicalPath,
            }));
            return r;
        }
    };
    return attributes_1.register(APIControllerType, serverContext, maishu_node_web_server_1.pathConcat(__filename, ""), { "list": ["/api/action/list", "/nws/actions"] });
}
exports.createAPIControllerType = createAPIControllerType;
