// maishu-wuzhui/out/Utility.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const maishu_toolkit_1 = require("maishu-toolkit");
class ElementHelper {
    static showElement(element) {
        if (!element)
            throw maishu_toolkit_1.errors.argumentNull('element');
        element.style.removeProperty('display');
    }
    static hideElement(element) {
        if (!element)
            throw maishu_toolkit_1.errors.argumentNull('element');
        element.style.display = 'none';
    }
    static isVisible(element) {
        let { display } = element.style;
        return !display || display != 'none';
    }
    static data(e, name, value) {
        let element = e;
        element['data'] = element['data'] || {};
        if (value == null)
            return element['data'].name;
        element['data'].name = value;
    }
    static findFirstParentByTagName(element, tagName) {
        if (element == null)
            throw maishu_toolkit_1.errors.argumentNull("element");
        if (!tagName)
            throw maishu_toolkit_1.errors.argumentNull('tagName');
        let parent = element.parentElement;
        while (parent != null) {
            if (parent.tagName.toLowerCase() == tagName.toLowerCase()) {
                return parent;
            }
            parent = parent.parentElement;
        }
        return null;
    }
}
exports.ElementHelper = ElementHelper;
function applyStyle(element, value) {
    if (typeof value == 'string') {
        element.setAttribute('style', value);
        return;
    }
    for (let key in value) {
        element.style[key] = value[key];
    }
}
exports.applyStyle = applyStyle;
