// maishu-jueying/out/editor-panel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EditorPanel = void 0;
const React = require("react");
const property_editor_1 = require("./property-editor");
const style_1 = require("./style");
class EditorPanel extends React.Component {
    constructor(props) {
        super(props);
        this.state = { componentDatas: [] };
    }
    /** 对输入进行验证 */
    validateInputs() {
        return this.editor.validator.checkAsync();
    }
    render() {
        let { empty } = this.props;
        empty = empty || React.createElement("div", { className: "empty" }, "\u6682\u65E0\u53EF\u7528\u7684\u5C5E\u6027");
        return React.createElement("div", { className: `${style_1.classNames.editorPanel} ${this.props.className || ""}`, ref: (e) => this.element = e || this.element },
            React.createElement(property_editor_1.PropertyEditor, { ref: e => this.editor = e || this.editor, empty: empty, customRender: this.props.customRender }));
    }
}
exports.EditorPanel = EditorPanel;
