// maishu-wuzhui-helper/out-es5/fields/checkbox-list-field.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _get(target, property, receiver) { if (typeof Reflect !== "undefined" && Reflect.get) { _get = Reflect.get; } else { _get = function _get(target, property, receiver) { var base = _superPropBase(target, property); if (!base) return; var desc = Object.getOwnPropertyDescriptor(base, property); if (desc.get) { return desc.get.call(receiver); } return desc.value; }; } return _get(target, property, receiver || target); }

function _superPropBase(object, property) { while (!Object.prototype.hasOwnProperty.call(object, property)) { object = _getPrototypeOf(object); if (object === null) break; } return object; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.checkboxListField = void 0;

var w = require("maishu-wuzhui");

function checkboxListField(params) {
  var field = new CheckboxListField(params);
  return field;
}

exports.checkboxListField = checkboxListField;

var CheckboxListField =
/*#__PURE__*/
function (_w$BoundField) {
  _inherits(CheckboxListField, _w$BoundField);

  function CheckboxListField(params) {
    var _this;

    _classCallCheck(this, CheckboxListField);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(CheckboxListField).call(this, params));
    _this.checkboxs = {};
    _this._params = params;
    return _this;
  }

  _createClass(CheckboxListField, [{
    key: "getDataItems",
    value: function getDataItems() {
      return __awaiter(this, void 0, void 0,
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee() {
        var r;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (this.dataItems) {
                  _context.next = 5;
                  break;
                }

                _context.next = 3;
                return this._params.dataSource.select();

              case 3:
                r = _context.sent;
                this.dataItems = r.dataItems;

              case 5:
                return _context.abrupt("return", this.dataItems);

              case 6:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));
    }
  }, {
    key: "setControlValue",
    value: function setControlValue(value) {
      value = value || [];

      for (var i = 0; i < value.length; i++) {
        var checkbox = this.checkboxs[value[i]];

        if (checkbox) {
          checkbox.checked = true;
        }
      }
    }
  }, {
    key: "createControl",
    value: function createControl() {
      var _this2 = this;

      var element = document.createElement("div");
      this.getDataItems().then(function (dataItems) {
        _this2.dataItems = dataItems;

        for (var i = 0; i < dataItems.length; i++) {
          console.assert(_this2._params.nameField != null);
          console.assert(_this2._params.valueField != null);
          var name = dataItems[i][_this2._params.nameField];
          var value = dataItems[i][_this2._params.valueField];
          var itemCtrl = document.createElement("div");
          itemCtrl.innerHTML = "<label><input type=\"checkbox\" value=\"".concat(value, "\">").concat(name, "</label>");
          itemCtrl.className = "checkbox";
          element.appendChild(itemCtrl);
          var checkbox = itemCtrl.querySelector("input");
          _this2.checkboxs["".concat(value)] = checkbox;
        }

        if (self.value) {
          _this2.setControlValue(self.value);
        }
      });
      var self = this;
      return {
        element: element,

        get value() {
          var checkboxs = [];

          for (var c in self.checkboxs) {
            if (self.checkboxs[c].checked) {
              checkboxs.push(self.checkboxs[c]);
            }
          }

          var value = checkboxs.map(function (o) {
            return o.value;
          });
          return value;
        },

        set value(value) {
          self.value = value;
          self.setControlValue(value);
        }

      };
    }
  }, {
    key: "createItemCell",
    value: function createItemCell(dataItem) {
      var field = this;

      var cell = _get(_getPrototypeOf(CheckboxListField.prototype), "createItemCell", this).call(this, dataItem);

      cell.render = function (dataItem) {
        var _this3 = this;

        field.getDataItems().then(function (dataItems) {
          var value = dataItem[field._params.dataField];
          var names = [];

          var _loop = function _loop(i) {
            var dataItem = dataItems.filter(function (o) {
              return o[field._params.valueField] == value[i];
            })[0];

            if (dataItem) {
              names.push(dataItem[field._params.nameField]);
            }
          };

          for (var i = 0; i < value.length; i++) {
            _loop(i);
          }

          var text = names.filter(function (o) {
            return o;
          }).join(",");
          _this3.element.innerHTML = text;
        });
      };

      return cell;
    }
  }]);

  return CheckboxListField;
}(w.BoundField);
